### PR TITLE
Download data and add database entry from HEK query

### LINKED ISSUE #848: Download data and add entries to database from a HEK query
This should also be handled by the `Database.download` method.

Notes:
- add a new method download_from_hek_query_result(query_result, path=None, progress=False) which translates the incoming HEK qr to a VSO qr and then calls download_from_vso_query_result and passes the parameters path and progress on to this method.

### PR DESCRIPTION
**Pull request by jains8844**
Fixes #848

### CODE REVIEW

**Comment by jains8844** [comment]
I tried mocking parfive.download but database entries are added from the headers of the files downloaded so files were required.

**Comment by jains8844** [comment]
Kindly review this

File: sunpy/database/database.py [hunk 1]
```diff
@@ -799,6 +799,42 @@ def add_from_hek_query_result(self, query_result,
             H2VClient().translate_and_query(query_result))
         self.add_from_vso_query_result(vso_qr, ignore_already_added)
 
+    def download_from_hek_query_result(self, query_result, client=None,
+                                       path=None, progress=False,
+                                       ignore_already_added=False, overwrite=False):
+        """download(query_result, client=sunpy.net.vso.VSOClient(),
+        path=None, progress=False, ignore_already_added=False)
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/database.py [hunk 2]
```diff
@@ -799,6 +799,42 @@ def add_from_hek_query_result(self, query_result,
             H2VClient().translate_and_query(query_result))
         self.add_from_vso_query_result(vso_qr, ignore_already_added)
 
+    def download_from_hek_query_result(self, query_result, client=None,
+                                       path=None, progress=False,
+                                       ignore_already_added=False, overwrite=False):
+        """download(query_result, client=sunpy.net.vso.VSOClient(),
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
        """
```

File: sunpy/database/database.py [hunk 3]
```diff
@@ -799,6 +799,42 @@ def add_from_hek_query_result(self, query_result,
             H2VClient().translate_and_query(query_result))
         self.add_from_vso_query_result(vso_qr, ignore_already_added)
 
+    def download_from_hek_query_result(self, query_result, client=None,
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
    def download_from_hek_query_result(self, query_result, client=None, path=None, progress=False,
```

File: sunpy/database/database.py [hunk 4]
```diff
@@ -799,6 +799,42 @@ def add_from_hek_query_result(self, query_result,
...
+                                       path=None, progress=False,
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/database.py [hunk 5]
```diff
@@ -799,6 +799,42 @@ def add_from_hek_query_result(self, query_result,
...
+                                       ignore_already_added=False, overwrite=False):
+        """download(query_result, client=sunpy.net.vso.VSOClient(),
+        path=None, progress=False, ignore_already_added=False)
+
+        Add new database entries from a hek query result by converting it
+        into vso query and download the corresponding data files.
+        See :meth:`sunpy.database.Database.download`
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/database.py [hunk 6]
```diff
@@ -799,6 +799,42 @@ def add_from_hek_query_result(self, query_result,
...
+        for information about the caching mechanism used and about the
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/database.py [hunk 7]
```diff
@@ -799,6 +799,42 @@ def add_from_hek_query_result(self, query_result,
...
+        parameters `client`, `path`, `progress`.
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/database.py [hunk 8]
```diff
@@ -799,6 +799,42 @@ def add_from_hek_query_result(self, query_result,
...
+
+        Parameters
+        ----------
+        query_result : HEKTable or HEKRow
+            The value returned by :meth:`sunpy.net.hek.HEKClient().search`
+
+        ignore_already_added : bool
+            See :meth:`sunpy.database.Database.add`.
+
```
Comment (nabobalis) [inline sunpy/database/database.py]: These parameters need to be expanded to cover all the input for this function.

File: sunpy/database/tests/test_database.py [hunk 9]
```diff
@@ -441,6 +441,26 @@ def test_add_entry_from_hek_qr(database):
...
+    )
+
+    database.download_from_hek_query_result(
+        query[4], path=str(tmpdir.join('{file}.fits')))
+
+    fits_file_pattern = str(tmpdir.join('*.fits'))
+    num_of_fits_headers = sum(
+        len(fits.get_header(file)) for file in glob.glob(fits_file_pattern))
+
+    assert len(database) > 0 and len(database) == num_of_fits_headers
```
Comment (nabobalis) [inline sunpy/database/tests/test_database.py]: How many files does this download and how large?

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: 10 files with total of around 100 megabytes

Comment (nabobalis) [reply sunpy/database/tests/test_database.py]: There is no way to make this smaller?

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: I tried many queries but couldn't find something smaller

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: One more method would be to keep the files in sample and mock parfive.download to return path to those files instead.

Comment (nabobalis) [reply sunpy/database/tests/test_database.py]: I don't think adding these files to sample data is the way forward here. I think somehow mocking the download and getting fake files would be better but I am unsure how to mock the download.

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: We need headers from the files since the database entries are made from them.

Comment (nabobalis) [reply sunpy/database/tests/test_database.py]: You mean fits headers? We can probably mock those too.

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: One more way would be to mock the function ``` entries_from_file ``` to return mock DatabaseEntry objects [here](https://github.com/sunpy/sunpy/blob/master/sunpy/database/database.py#L419).

Comment (nabobalis) [reply sunpy/database/tests/test_database.py]: We also don't want to download any files if we can help it or at least not that many.

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: The above method won't download any file. But we can't really test then if the files are downloaded or not.

Comment (nabobalis) [reply sunpy/database/tests/test_database.py]: Whatever works to lessen downloading 10 files.

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: So should I do the ```entries_from_file``` mocking thing ?

Comment (nabobalis) [reply sunpy/database/tests/test_database.py]: If that will work.

**Comment by jains8844** [comment]
I've updated the tests accordingly

File: sunpy/database/database.py [hunk 10]
```diff
@@ -799,6 +799,53 @@ def add_from_hek_query_result(self, query_result,
...
+        path : `str`
+            Path to download the files.
+
+        progress : `bool`
+            If True, displays the progress bar during file download.
+
+        ignore_already_added : `bool`
+            See :meth:`sunpy.database.Database.add`.
+
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/database.py [hunk 11]
```diff
@@ -799,6 +799,53 @@ def add_from_hek_query_result(self, query_result,
...
+        client : `sunpy.net.vso.VSOClient`, optional
+            VSO Client instance to use for search and download.
+            If not specified a new instance will be created.
+
+        path : `str`
+            Path to download the files.
+
+        progress : `bool`
+            If True, displays the progress bar during file download.
+
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/database.py [hunk 12]
```diff
@@ -799,6 +799,53 @@ def add_from_hek_query_result(self, query_result,
...
+        Parameters
+        ----------
+        query_result : `HEKTable` or `HEKRow`
+            The value returned by :meth:`sunpy.net.hek.HEKClient().search`
+
+        client : `sunpy.net.vso.VSOClient`, optional
+            VSO Client instance to use for search and download.
+            If not specified a new instance will be created.
+
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/database.py [hunk 13]
```diff
@@ -799,6 +799,53 @@ def add_from_hek_query_result(self, query_result,
...
+                                       ignore_already_added=False, overwrite=False):
+        """
+        Add new database entries from a hek query result by converting it
+        into vso query and download the corresponding data files.
+
+        Parameters
+        ----------
+        query_result : `HEKTable` or `HEKRow`
+            The value returned by :meth:`sunpy.net.hek.HEKClient().search`
+
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/database.py [hunk 14]
```diff
@@ -799,6 +799,53 @@ def add_from_hek_query_result(self, query_result,
...
+        client : `sunpy.net.vso.VSOClient`, optional
+            VSO Client instance to use for search and download.
+            If not specified a new instance will be created.
+
+        path : `str`
+            Path to download the files.
+
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

File: sunpy/database/tests/test_database.py [hunk 15]
```diff
@@ -441,6 +445,52 @@ def test_add_entry_from_hek_qr(database):
     assert len(database) > 1
 
 
+@pytest.mark.remote_data
```
Comment (nabobalis) [inline sunpy/database/tests/test_database.py]: Do we still need this?

File: sunpy/database/tests/test_database.py [hunk 16]
```diff
@@ -441,6 +445,52 @@ def test_add_entry_from_hek_qr(database):
...
+    monkeypatch.setattr(tables, "entries_from_dir", mock_entries_from_dir)
+
+    query = hek.HEKClient().search(
+        hek.attrs.Time('2019/03/10 14:40:10', '2019/04/11 16:40:50'),
+        hek.attrs.EventType('FL')
+    )
+
+    database.download_from_hek_query_result(query[4], path=str(tmpdir))
+
+    assert len(database) > 0 and len(database) == 10
```
Comment (nabobalis) [inline sunpy/database/tests/test_database.py]: Having both checks seems redundant?

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: You mean checking queue length and queue data or checking length of database at the end ?

File: sunpy/database/tests/test_database.py (hunk 15 above)
Comment (jains8844) [reply sunpy/database/tests/test_database.py]: Yes this is for search

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: And also for h2vclient used to convert hek query to vso query

File: sunpy/database/tests/test_database.py (hunk 16 above)
Comment (nabobalis) [reply sunpy/database/tests/test_database.py]: Yeah.

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: So should I remove the check for length of database ?

Comment (nabobalis) [reply sunpy/database/tests/test_database.py]: Just have one of them?

Comment (jains8844) [reply sunpy/database/tests/test_database.py]: Yeah sure :`)

**Comment by jains8844** [comment]
Is there anything left to be done in this ??

**Comment by nabobalis** [comment]
It will need to be reviewed by two people first and then you can respond to any comments that come up but in the meantime there is nothing else to do from you but to wait. You can move onto other things or issues as you please.

File: sunpy/database/database.py [hunk 17]
```diff
@@ -799,6 +799,48 @@ def add_from_hek_query_result(self, query_result,
...
+            If True, displays the progress bar during file download.
+        ignore_already_added : `bool`
+            See :meth:`sunpy.database.Database.add`.
+        overwrite : `bool`, optional
+            If True, matching database entries from the query results will be
+            deleted and replaced with new database entries, with all files
+            getting downloaded.
+            Otherwise, no new file download and update of matching database
+            entries takes place.
+
```
Comment (nabobalis) [inline sunpy/database/database.py]:
```suggestion
```

**Comment by dstansby** [comment]
Could you merge in the current master branch to fix the doc build?

**Comment by jains8844** [comment]
> Could you merge in the current master branch to fix the doc build?

So should I open a different PR from the master branch ??

**Comment by nabobalis** [comment]
> > Could you merge in the current master branch to fix the doc build?
> 
> So should I open a different PR from the master branch ??

No. Something like this: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-an-upstream-repository-into-your-fork

**Comment by jains8844** [comment]
Oh I have to update the fork. I thought something different :`)

**Comment by jains8844** [comment]
Kindly tell if anything is left to be done in this PR ??

**Comment by nabobalis** [comment]
No there is nothing more for you to do other than to wait for one more person to review this pull request.

**Pull request merged**